Improve performance of Time to/from JSON

bboreham opened the pull request:
A few changes to speed up marshalling/unmarshalling of `model.Time` - see individual commit messages.

Benchmarks: before:
```
BenchmarkMarshalJSON-2     	 5000000	       303 ns/op	      64 B/op	       3 allocs/op
BenchmarkUnmarshalJSON-2   	 5000000	       314 ns/op	      53 B/op	       5 allocs/op
```

after:
```
BenchmarkMarshalJSON-2     	10000000	       219 ns/op	      16 B/op	       1 allocs/op
BenchmarkUnmarshalJSON-2   	10000000	       124 ns/op	      16 B/op	       2 allocs/op
```

@fabxc

bboreham commented:
Note prometheus/prometheus#3536 overrides `MarshalJSON` in `Point` so this PR has less impact.  I still have cases where I'm marshalling `Time` directly.

#3536 doesn't touch the `UnmarshalJSON` side.

brian-brazil commented:
Where is this code getting used? We're trying to move away from model generally.

bboreham commented:
Aha, hadn't picked up on that shift.

Used here: https://github.com/weaveworks/cortex/blob/cc8cdf499162077c4a21ae05d162ca8116228718/pkg/chunk/chunk.go#L44-L45
no biggy to change it to underlying types.

brian-brazil commented:
Changing to underlying types is probably best, it'll be less of a problem when this code eventually goes away.

bboreham commented:
In the end I moved my code to use `jsoniter` and added type overrides to achieve the same effect.

You might consider a note in the README or CONTRIBUTING here that `model` is deprecated.
